File: rust/src/contains_duplicate.rs
//! This module was written to solve the "Contains Duplicate" problem from LeetCode.
//!
//! Given an integer array `nums` and an integer `k, return `true` if there are two distinct indices `i` and `j` in the array such that `nums[i] == nums[j]` and `abs(i - j) <= k`.
//! 
//! Examples:
//! Input: nums = [1,2,3,1], k = 3
//! Output: true
//! 
//! Input: nums = [1,2,3,1,2,3], k = 2
//! Output: false
//! 
//! Usage: contains_nearby_duplicate(nums, k)

use std::collections::HashMap;

/// Determines if there is a number with a duplicate within `k` range
pub fn contains_nearby_duplicate(nums: Vec<i32>, k: i32) -> bool {
  let mut last_indexes: HashMap<i32, u32> = HashMap::new();

  for (i, num) in nums.into_iter().enumerate() {
    let idx = i as u32;
    match last_indexes.get(&num) {
      Some(v) => {
        if i32::abs((idx - v) as i32) <= k {
            return true;
        }
        last_indexes.insert(num, idx);
      }
      None => {
        last_indexes.insert(num, idx);
      }
    }
  }

  false
}

File: py/contains_duplicate.py
"""
Given an integer array nums, return true if any value appears at least twice in the array, and return false if every element is distinct.

Examples:
Input: nums = [1,2,3,1]
Output: true

Input: nums = [1,2,3,4]
Output: false

Input: nums = [1,1,1,3,3,4,3,2,4,2]
Output: true

Problem source: LeetCode

Usage: contains_duplicate.py <comma_separated_nums>
"""

from typing import List
from utils import read_int_array
import sys

def contains_duplicate(nums: List[int]) -> bool:
  counts = set()

  for num in nums:
    if num not in counts:
      counts.add(num)
    else:
      return True

  return False


if __name__ == "__main__":
  print(contains_duplicate(read_int_array(sys.argv[1])))